envision/server.py

```diff
@@ -214,28 +220,38 @@ def sync_run_forever():
         self._thread = threading.Thread(target=sync_run_forever, args=(), daemon=True)
         self._thread.start()
 
-    def _push_frame_to_web_client(self, frame):
+    def _push_frames_to_web_client(self, frames):
         try:
-            self._client.write_message(
-                json.dumps(
-                    {
-                        "state": frame.data,
-                        "current_elapsed_time": frame.timestamp
-                        - self._frames.start_time,
-                        "total_elapsed_time": self._frames.elapsed_time,
-                    }
-                )
-            )
+            frames_formatted = [
+                {
+                    "state": frame.data,
+                    "current_elapsed_time": frame.timestamp - self._frames.start_time,
+                    "total_elapsed_time": self._frames.elapsed_time,
+                }
+                for frame in frames
+            ]
+            self._client.write_message(json.dumps(frames_formatted))
             return False
         except WebSocketClosedError:
             return True
 
-    async def _wait_for_next_frame(self, frame):
+    def _calculate_frame_delay(self, frame_ptr):
+        return 0.5 * self._timestep_sec if not frame_ptr.next_ else 0
+
+    async def _wait_for_next_frame(self, frame_ptr):
         while True:
-            # TODO: Consider using an asyncio queue instead
-            await asyncio.sleep(self._timestep_sec)
-            if frame.next_ is not None:
-                return frame.next_
+            delay = self._calculate_frame_delay(frame_ptr)
+            await asyncio.sleep(delay)
+
+            if frame_ptr.next_ is not None:
+                frames_to_send = []
+                count = 0
```

Comment:
`count` isn't really needed as a variable.  you can just use `len(frames_to_send)`.